File: obfuscate.py
import random 
import string

# Global Variables
random_values = []

# Generate Random String of 5-8 Characters
def generate_random_string():
    global random_values
    while [ True ]:
        rand = "".join(random.choice(string.ascii_letters) for _ in range(random.randint(5,9)))
        if rand not in random_values:
            random_values.append(rand)
            return rand

# Character --> String Map
def map_characters(cmd, space_token, equals_token):
    mapping = {}
    mapping[" "] = space_token
    mapping["="] = equals_token
    for part in cmd:
        if part not in mapping:
            mapping[part] = generate_random_string()
    return mapping

# Payload Object
class Create_Payload:
    def __init__(self, cmd):
        self.cmd = cmd
        self.bad_char = string.punctuation
        self.set_token = generate_random_string()
        self.space_token = generate_random_string()
        self.equals_token = generate_random_string()
        self.mapping = map_characters(self.cmd, self.space_token, self.equals_token)

    # Initial Setup for Obfuscation
    def initial_setup(self):
        initial = [
            f"set {self.set_token}=set",
            f"%{self.set_token}% {self.space_token}= ",
            f"%{self.set_token}%%{self.space_token}%{self.equals_token}=="
        ]
        return initial

    # Use Character Mapping to Craft Payload (After Initial Setup)
    def obfuscate(self, initial_payload):
        new_payload = []
        command_parts = ""
        for part in self.cmd:
            if part in self.bad_char:
                command_parts += part                
            else:
                obfuscated_part = self.mapping[part]
                command_parts += f"%{obfuscated_part}%"
                new_payload += [
                    f"%{self.set_token}%%{self.space_token}%{obfuscated_part}%{self.equals_token}%{part}"
                ]
        random.shuffle(new_payload)
        return "{}\n{}{}".format("\n".join(initial_payload), "&".join(new_payload), "\n{}".format(command_parts))

def main():
    # Create Payload Object w/ Command to Obfuscate
    command = input("Command> ").strip()
    payload = Create_Payload(command)
    initial_payload = payload.initial_setup()
    obfuscated_payload = payload.obfuscate(initial_payload)

    # Write to .bat File
    print("\n[+] Command: {}\n[+] Payload Size: {} Characters\n[+] Payload:\n{}\n\n[+] Writing to payload.bat...".format(command, len(obfuscated_payload), obfuscated_payload))
    with open("payload.bat", "w") as file_handle:
        try:
            file_handle.write(obfuscated_payload)
            print("[+] Payload Successfully Written!")
        except:
            print("[-] Unknown File Error has Occured.")

if __name__ == "__main__":
    main()

File: obfuscate.go
package main

import (
	"fmt"
	"math/rand"
	"time"
	"bufio"
	"os"
	"strings"
	"log"
)

var (
	letters = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
)

// Construct Obfucated Payload func
type obfuscate func(string, map[string]string) string

// Struct to Build Payload
type payload struct {
	cmd string
	mapping map[string]string
	obfuscated obfuscate
}

func main() {
	// Seed Random
	rand.Seed(time.Now().UnixNano())

	// Map Characters, Get Input, Create Payload
	mapping := mapChars()
	reader := bufio.NewReader(os.Stdin)
	fmt.Print("Command> ")
	cmd, _ := reader.ReadString('\n')
	cmd = strings.ReplaceAll(cmd, "\r\n", "")
	cmd = strings.ReplaceAll(cmd, "\n", "")
	p := payload{
		cmd: cmd,
		mapping: mapping,
		obfuscated: func(cmd string, mapping map[string]string) string {
			var initial, getLetters, command []string
			var initialPart, letterPart, commandPart string

			// Initial Setup
			initial = append(initial, fmt.Sprintf("set %s=set", mapping["set"]))
			initial = append(initial, fmt.Sprintf("%%%s%% %s= ", mapping["set"], mapping[" "]))
			initial = append(initial, fmt.Sprintf("%%%s%%%%%s%%%s==", mapping["set"], mapping[" "], mapping["="]))
			initialPart =  strings.Join(initial, "\n")

			// Letters Setup
			for i := range letters {
				if strings.Contains(cmd, string(letters[i])){
					getLetters = append(getLetters, fmt.Sprintf("%%%s%%%%%s%%%s%%%s%%%s", mapping["set"],mapping[" "], string(letters[i]),mapping["="], mapping[string(letters[i])]))
				}
			}
			letterPart = strings.Join(getLetters, " & ")

			// Command Setup
			for i := range cmd {
				command = append(command, fmt.Sprintf("%%%s%%", mapping[string(cmd[i])]))
			}
			commandPart = strings.Join(command, "")

			// Combine Initial + Letters + Command
			return fmt.Sprintf("%s\n%s\n%s", initialPart, letterPart, commandPart)
		},
	}

	// Constructed Payload
	constructed := p.obfuscated(p.cmd, p.mapping)

	// Output Information
	fmt.Printf("\n[+] Command: %s\n[+] Payload Size: %d Characters\n[+] Payload:\n%s\n\n[+] Writing to payload.bat...", p.cmd, len(constructed), constructed)

	// Write to a File
	f, err := os.Create("payload.bat")
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	_, err = f.WriteString(constructed)
	if err != nil {
		log.Fatal(err)
	}

}

// Take Length, Return Random String of Given Length
func genRandomString(length int) string {
	str := make([]rune, length)
	for i := range str {
		// Choose Random Letters
		str[i] = letters[rand.Intn(len(letters))]
	}
	return string(str)
}

// Map Characters A-Z, a-z, "=", " " and "set" to Unique Random Strings 
func mapChars() map[string]string {
	mapping := make(map[string]string)
	var randNum int
	var attempt string
	// A-Z, a-z
	for i := range letters {
		randNum = rand.Intn(6) + 4
		attempt = genRandomString(randNum)
		for {
			// Check if Value Exists
			if uniq := checkUniq(mapping, attempt); uniq {
				randNum = rand.Intn(6) + 4
				attempt = genRandomString(randNum)
				mapping[string(letters[i])] = attempt
				break
			}
		}
	}
	// Get Random String for "=", " ", and "set"
	other := [3]string{"=", " ", "set"}
	for i := range other {
		randNum = rand.Intn(6) + 4
		attempt = genRandomString(randNum)
		for {
			// Check if Value Exists
			if uniq := checkUniq(mapping, attempt); uniq {
				randNum = rand.Intn(6) + 4
				attempt = genRandomString(randNum)
				mapping[other[i]] = attempt
				break
			}
		}
	}

	return mapping
}

// Function to Check if Value is Unique in a Map
func checkUniq(mapping map[string]string, value string) bool {
	for i := range letters {
		if val, ok := mapping[string(letters[i])]; ok {
			if val == value {
				return false
			}
		}
	}
	return true
}
